### PR TITLE
feat($table): add CourseSearch table filter and paignation

### PR DESCRIPTION
**Pull request by BinnyDaBin**

### CODE REVIEW

File: src/components/table/Table.js [hunk 1]
```diff
@@ -1,28 +1,417 @@
...
+        ) : (
+          <KeyboardArrowRight />
+        )}
+      </IconButton>
+      <IconButton
+        onClick={handleLastPageButtonClick}
+        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
+        aria-label='Last Page'
+      >
+        {theme.direction === 'rtl' ? <FirstPageIcon /> : <LastPageIcon />}
```
Comment (kebabcase) [inline src/components/table/Table.js]: These could be functions

```js
const getPageIcon = (type, direction) => {
  if (type === 'first') {
    return direction === 'rtl' ? <LastPageIcon /> : <FirstPageIcon />;
  }
  if (type === 'last') {
    return direction === 'rtl' ? <FirstPageIcon /> : <LastPageIcon />;
  }
};


...

return (
  ...
  <IconButton>
    {getPageIcon('first', theme.direction)}
  </IconButton>
  ...
);
```

File: src/components/table/Table.js [hunk 2]
```diff
@@ -1,28 +1,417 @@
...
+        ) : (
+          <KeyboardArrowLeft />
+        )}
+      </IconButton>
+      <IconButton
+        onClick={handleNextButtonClick}
+        disabled={page >= Math.ceil(count / rowsPerPage) - 1}
+        aria-label='Next Page'
+      >
+        {theme.direction === 'rtl' ? (
```
Comment (kebabcase) [inline src/components/table/Table.js]: Same comment as below `getPageIcon()` suggestion

File: src/components/table/Table.js [hunk 3]
```diff
@@ -1,28 +1,417 @@
...
+  rowsPerPage,
+  totalRows,
+  onPaginate,
+  filterSelects,
+  onFilter
+}) => {
+  const tableClasses = useTableStyles();
+  const FilterClasses = useFilterStyles();
+
+  const [filters, setFilters] = useState({
```
Comment (kebabcase) [inline src/components/table/Table.js]: Why not

```js
const [filters, setFilters] = useState(_.transform(colConfig, (initialFilter, value, key) => {
  initialFilter[key] = [];
  return initialFilter;
}, {}));
```

File: src/components/table/Table.js [hunk 4]
```diff
@@ -1,28 +1,417 @@
...
+    kzooCourseName: [],
+    kzooDiscipline: [],
+    hostInstiCourseNumber: [],
+    hostInstiCourseName: []
+  });
+
+  // call filter function everytime filters state gets updated
+  useEffect(() => {
+    onFilter(filters);
+  }, [onFilter, filters]);
```
Comment (kebabcase) [inline src/components/table/Table.js]: Not sure if `onFilter` is needed in this array param...? Isn't `useEffect` only get invoked when `filters` change?

File: src/components/table/Table.js [hunk 5]
```diff
@@ -1,28 +1,417 @@
...
+    filterSelects.studyAbroadInstitutionColumnOptions;
+  const programNames = filterSelects.programNameColumnOptions;
+  const years = filterSelects.yearColumnOptions;
+  const kzooCourseNames = filterSelects.kzooCourseNameColumnOptions;
+  const kzooDisciplines = filterSelects.kzooDisciplineColumnOptions;
+  const hostInstiCourseNumbers =
+    filterSelects.hostInstiCourseNumberColumnOptions;
+  const hostInstiCourseNames = filterSelects.hostInstiCourseNamesColumnOptions;
+
+  // Pagination Change
```
Comment (kebabcase) [inline src/components/table/Table.js]: Unnecessary comment

File: src/components/table/Table.js [hunk 6]
```diff
@@ -1,28 +1,417 @@
...
+  const handleChangePage = newPage => {
+    onPaginate(newPage);
+  };
+
+  // Filter Change
```
Comment (kebabcase) [inline src/components/table/Table.js]: Unnecessary comment

File: src/components/table/Table.js [hunk 7]
```diff
@@ -1,28 +1,417 @@
...
-        <th key={index}>{col.display}</th>
+    <TableRow>
+      {colConfig.map((column, index) => (
+        <TableCell key={index}>
+          <div className={FilterClasses.root}>
+            <FormControl className={FilterClasses.formControl}>
+              <InputLabel shrink htmlFor='select-multiple-checkbox'>
+                {column.display}
+              </InputLabel>
+              {column.id === 'studyAbroadInstitution' && (
```
Comment (kebabcase) [inline src/components/table/Table.js]:
```js
<Select
  multiple
  value={filters[column.id]}
  onChange={event => handleFilterChange(event, column.id)}
  input={<Input id='select-multiple-checkbox' />}
  renderValue={selected => selected.join(', ')}
  MenuProps={MenuProps}
>
  {
    filterSelects[column.id].map(selectOption => {
      <MenuItem key={selectOption} value={selectOption}>
        <Checkbox  checked={filters[column.id]indexOf(selectOption) > -1} />
        <ListItemText primary={institution} />
      </MenuItem>
    });
  }
</Select>
```

File: src/components/courses/CourseSearch.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+      size: coursesPerPage
+    };
+
+    getCourses(params);
+  }, [filters, currentPage, coursesPerPage]);
+
+  const getCourses = async ({ filters, offset, size }) => {
+    // setLoading(true);
+
+    const coursesResponse = await Requestor.get(
```
[the remark this answers is not in the document]
Comment (kebabcase) [reply src/components/courses/CourseSearch.js]:
```js
{
  filters: _.filter(filters, (filterValue) => {
    return !_.isEmpty(filterValue);
  }),
}
```

[the remark this answers is not in the document]
Comment (BinnyDaBin) [reply src/components/courses/CourseSearch.js]: _.filter returns an array with the accepted values from the object. In order to get an object with key/value pairs, use _.pickBy() instead

[the remark this answers is not in the document]
Comment (BinnyDaBin) [reply src/components/courses/CourseSearch.js]: Replace filters param to 
filters: _.pickBy(filters, filter => filter.length > 0)

File: src/components/table/Table.js (hunk 3 above)
Comment (BinnyDaBin) [reply src/components/table/Table.js]: Fix this to:
const [filters, setFilters] = useState(
    _.transform(
      colConfig,
      (initialFilter, config) => {
        initialFilter[config.id] = [];
        return initialFilter;
      },
      {}
    )
  );

File: src/components/table/Table.js (hunk 4 above)
Comment (BinnyDaBin) [reply src/components/table/Table.js]: Deleting onFilter from the dependency invokes this warning:

> Line 173:  React Hook useEffect has a missing dependency: 'onFilter'. Either include it or remove the dependency array. If 'onFilter' changes too often, find the parent component that defines it and wrap that definition in useCallback  react-hooks/exhaustive-deps

**Comment by kebabcase** [comment]
:+1: :ship:

**Pull request merged**